CloudMonitoring: Add support for preprocessing (#33011)
grafana/grafana: docs/sources/datasources/google-cloud-monitoring/_index.md

When the operator is set to `=~` or `!=~` it is possible to add regular expressions to the filter value field. E.g `us-central[1-3]-[af]` would match all values that starts with "us-central", is followed by a number in the range of 1 to 3, a dash and then either an "a" or an "f". Leading and trailing slashes are not needed when creating regular expressions.

#### Pre-processing

Preprocessing options are displayed in the UI when the selected metric has a metric kind of `delta` or `cumulative`. If the selected metric has a metric kind of `gauge`, no pre-processing option will be displayed.

If you select 'Rate', data points are aligned and converted to a rate per time series. If you select 'Delta', data points are aligned by their delta (difference) per time series.

#### Grouping

You can reduce the amount of data returned for a metric by combining different time series. To combine multiple time series, you can specify a grouping and a function.

##### Group by

Group by resource or metric labels to reduce the number of time series and to aggregate the results by a group by. E.g. Group by instance_name to see an aggregated metric for a Compute instance.

###### Metadata labels

> Resource metadata labels contain information to uniquely identify a resource in Google Cloud. Metadata labels are only returned in the time series response if they're part of the **Group By** segment in the time series request. There's no API for retrieving metadata labels, so it's not possible to populate the group by dropdown with the metadata labels that are available for the selected service and metric. However, the **Group By** field dropdown comes with a pre-defined list of common system labels.

Reviewer: ```suggestion
Resource metadata labels contain information to uniquely identify a resource in Google Cloud. Metadata labels are only returned in the time series response if they're part of the **Group By** segment in the time series request.

There's no API for retrieving metadata labels, so you can't populate the group by list with the metadata labels that are available for the selected service and metric. However, the **Group By** field list comes with a pre-defined set of common system labels.
```

> User labels cannot be pre-defined, but it's possible to enter them manually in the **Group By** field. If a metadata label, user label or system label is included in the **Group By** segment, then you can create filters based on it and expand its value on the **Alias** field.

Reviewer: ```suggestion
User labels cannot be pre-defined, but you can enter them manually in the **Group By** field. If a metadata label, user label or system label is included in the **Group By** segment, then you can create filters based on it and expand its value on the **Alias** field.
```

##### Group by function

> The grouping function is used to combine the time series in the group into a single time series.

Reviewer: ```suggestion
Use the grouping function to combine the time series in the group into a single time series.
```

Reviewer: How do you use it?

Author: Good point! I've updated the text

#### Alignment

The process of alignment consists of collecting all data points received in a fixed length of time, applying a function to combine those data points, and assigning a timestamp to the result.

##### Alignment function

> Describes how to bring the data points in each individual time series into equal periods of time. Within each interval (determined by the alignment period) and for each time series, the data points are aggregated into a single point. The value of that point is determined by the type of alignment function used. For more information about aligner, refer to [alignment metric selector](https://cloud.google.com/monitoring/charts/metrics-selector#alignment).

Reviewer: ```suggestion
Describes how to bring the data points in each individual time series into equal periods of time. Within each interval (determined by the alignment period) and for each time series, the data points are aggregated into a single point. The value of that point is determined by the type of alignment function used. For more information about alignment functions, refer to [alignment metric selector](https://cloud.google.com/monitoring/charts/metrics-selector#alignment).
```
```suggestion
Describes how to bring the data points in each individual time series into equal periods of time. Within each interval (determined by the alignment period) and for each time series, the data points are aggregated into a single point. The value of that point is determined by the type of alignment function used. For more information about aligner, refer to [alignment metric selector](https://cloud.google.com/monitoring/charts/metrics-selector#alignment).
```

##### Group by

Group by resource or metric labels to reduce the number of time series and to aggregate the results by a group by. E.g. Group by instance_name to see an aggregated metric for a Compute instance.

###### Metadata labels

Resource metadata labels contain information to uniquely identify a resource in Google Cloud. Metadata labels are only returned in the time series response if they're part of the **Group By** segment in the time series request.

There's no API for retrieving metadata labels, so you can't populate the group by list with the metadata labels that are available for the selected service and metric. However, the **Group By** field list comes with a pre-defined set of common system labels.

User labels cannot be pre-defined, but you can enter them manually in the **Group By** field. If a metadata label, user label or system label is included in the **Group By** segment, then you can create filters based on it and expand its value on the **Alias** field.

##### Group by function

Select a function in the grouping function select to combine the time series in the group into a single time series.

#### Alignment

> The process of alignment consists of collecting all data points received in a fixed length of time, applying a function to combine those data points, and assigning a timestamp to the result.

Author: Not sure about the changes before the first comma. What about, `During alignment, all data data points...`?

> The process of alignment consists of collecting all data points received in a fixed length of time, applying a function to combine those data points, and assigning a timestamp to the result.

Reviewer 2: Makes sense.

##### Alignment function

> Describes how to bring the data points in each individual time series into equal periods of time. Within each interval (determined by the alignment period) and for each time series, the data points are aggregated into a single point. The value of that point is determined by the type of alignment function used. For more information about alignment functions, refer to [alignment metric selector](https://cloud.google.com/monitoring/charts/metrics-selector#alignment).

Reviewer 2: ```suggestion
Alignment brings the data points in each time series into equal intervals. Within each interval (determined by the alignment period) and for each time series, the data is aggregated into a single point. The value of that point is determined by the type of alignment function used. For more information on alignment functions, refer to [alignment metric selector](https://cloud.google.com/monitoring/charts/metrics-selector#alignment).
```

> It is also possible to choose fixed time intervals to group by, like `1h` or `1d`.

Reviewer: ```suggestion
You can also choose fixed time intervals to group by, like `1h` or `1d`.
```
